### PR TITLE
paper-plane: init at 0.1.0-beta.5

### LINKED ISSUE #281095: Package request: Paper Plane
**Project description**

Paper Plane is an alternative Telegram client. It uses libadwaita for its user interface and strives to meet the design principles of the GNOME desktop.

Keep in mind Paper Plane is still in Beta releases.

**Metadata**

* homepage URL: https://github.com/paper-plane-developers/paper-plane
* source URL: https://github.com/paper-plane-developers/paper-plane
* license: GPL-3.0
* platforms: Linux

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by Aleksanaa**
## Description of changes

Darwin should be supported but not tested yet.

Closes https://github.com/NixOS/nixpkgs/issues/281095

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Aleksanaa** [comment]
```
  /private/tmp/nix-build-paper-plane-0.1.0-beta.5.drv-0/tfacd-0/gettext/libtextstyle/lib/obstack.c:351:31: error: incompatible function pointer types initializing 'void (*)(void) __attribute__((noreturn))' with an expression of type 'void (void)' [-Wincompatible-function-pointer-types]
  __attribute_noreturn__ void (*obstack_alloc_failed_handler) (void)
                                ^
  1 error generated.
  make[5]: *** [Makefile:2448: obstack.lo] Error 1
  make[5]: *** Waiting for unfinished jobs....
  make[4]: *** [Makefile:2157: all] Error 2
  make[3]: *** [Makefile:1786: all-recursive] Error 1
  make[2]: *** [Makefile:1714: all] Error 2
  make[1]: *** [Makefile:403: all-recursive] Error 1
  make: *** [Makefile:359: all] Error 2
  thread 'main' panicked at /private/tmp/nix-build-paper-plane-0.1.0-beta.5.drv-0/cargo-vendor-dir/gettext-sys-0.21.3/build.rs:300:5:

  command did not execute successfully, got: exit status: 2

  build script failed, must exit now
```

It seems to be gettext-sys, but I'm not sure why

**Comment by Aleksanaa** [comment]
https://github.com/Koka/gettext-rs/issues/114

**Comment by Aleksanaa** [comment]
@Guanran928 Please test again if you have time, thanks!

**Comment by ocfox** [comment]
Result of `nixpkgs-review pr 284804` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>paper-plane</li>
  </ul>
</details>

**Comment by Aleksanaa** [comment]
> Result of `nixpkgs-review pr 284804` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
> 1 package failed to build:
> 
>     * paper-plane

Now I'd guess what happened on your machine because I can't afford a macOS device ;(

**Comment by ocfox** [comment]
Possibly forgot `darwin.apple_sdk.frameworks`.

**Comment by Aleksanaa** [comment]
> Possibly forgot `darwin.apple_sdk.frameworks`.

I know for sure, but what's next?

**Comment by Aleksanaa** [comment]
Should fix it. Linux doesn't seem to select the same gtk4 library from two directories at the same time (one is the patched gtk4, the other is the original, probably introduced due to libadwaita), so no problem occurs.

**Comment by Aleksanaa** [comment]
Should fix it. Linux doesn't seem to select the same gtk4 library from two directories at the same time (one is the patched gtk4, the other is the original, probably introduced due to libadwaita), so no problem occurs.

**Comment by Aleksanaa** [comment]
> However the icon in the login page isn't showing properly...

Does this also happen for other gtk4 apps?

**Comment by Aleksanaa** [comment]
I'm not sure if I should include an icon theme in the package closure, as this may not be desired. See related discussion here: https://github.com/NixOS/nixpkgs/pull/43150#issuecomment-450058333
Also libadwaita doesn't pick up a default icon theme. That's strange.

**Comment by Aleksanaa** [comment]
You can try `adwaita-1-demo` in `libadwaita.devdoc`.

**Comment by Aleksanaa** [comment]
Then that's out of scope for this pr. An icon theme should be installed to make it display normally.

File: pkgs/by-name/pa/paper-plane/rlottie.nix
Comment (wegank) [inline pkgs/by-name/pa/paper-plane/rlottie.nix:4]:
```suggestion
```

File: pkgs/by-name/pa/paper-plane/rlottie.nix
Comment (wegank) [inline pkgs/by-name/pa/paper-plane/rlottie.nix:10]:
```suggestion
stdenv.mkDerivation {
```

File: pkgs/by-name/pa/paper-plane/tdlib.nix
Comment (wegank) [inline pkgs/by-name/pa/paper-plane/tdlib.nix:3]:
```suggestion
stdenv.mkDerivation {
```

File: pkgs/by-name/pa/paper-plane/rlottie.nix
Comment (wegank) [inline pkgs/by-name/pa/paper-plane/rlottie.nix:12]:
```suggestion
  version = "0-unstable-2022-09-14";
```

https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming

File: pkgs/by-name/pa/paper-plane/rlottie.nix
Comment (wegank) [inline pkgs/by-name/pa/paper-plane/rlottie.nix:18]:
```suggestion
    hash = "sha256-OIKnDikuJuRIR9Jvl1PnUA9UAV09EmgGdDTeWoVi7jk=";
```

File: pkgs/by-name/pa/paper-plane/rlottie.nix
Comment (wegank) [inline pkgs/by-name/pa/paper-plane/rlottie.nix:27]: No maintainers?

File: pkgs/by-name/pa/paper-plane/package.nix
Comment (wegank) [inline pkgs/by-name/pa/paper-plane/package.nix:95]:
```suggestion
      lib.versionAtLeast stdenv.cc.version "16"
```

**Review by wegank — APPROVED** [review]

File: pkgs/by-name/pa/paper-plane/tdlib.nix
Comment (kirillrdy) [inline pkgs/by-name/pa/paper-plane/tdlib.nix:24]: replace-fail is proffered

Comment (Aleksanaa) [reply pkgs/by-name/pa/paper-plane/tdlib.nix:24]: It's just an old revision of tdlib copied from git history, don't expect too much from it

Comment (kirillrdy) [reply pkgs/by-name/pa/paper-plane/tdlib.nix:24]: `replace-fail` is really optional as its very recent change, i would't want to hold this PR up

**Comment by Aleksanaa** [comment]
@wegank able to merge?

File: pkgs/by-name/pa/paper-plane/tdlib.nix
Comment (kirillrdy) [inline pkgs/by-name/pa/paper-plane/tdlib.nix:1]: this looks like a copy and paste of ./pkgs/development/libraries/tdlib/default.nix

why not just override the version ?

Comment (Aleksanaa) [reply pkgs/by-name/pa/paper-plane/tdlib.nix:1]: This is probably not a good idea. For example, workarounds may be added after tdlib is updated, but these operations are not applicable to older versions, causing an increase in maintenance costs.

Comment (kirillrdy) [reply pkgs/by-name/pa/paper-plane/tdlib.nix:1]: one could argue same for duplication, there can be a change in dependency eg `gperf openssl readline zlib` or `gcc` that will require fixes to both copies of tdlib

maybe others can weigh in ? cc @SuperSandro2000 @mweinelt

Comment (Aleksanaa) [reply pkgs/by-name/pa/paper-plane/tdlib.nix:1]: The dependencies of the new version are not necessarily related to the dependencies of the old version. If we need to reuse these dependencies, more problems may be introduced. Likewise, meta.licenses are not necessarily consistent.

Comment (kirillrdy) [reply pkgs/by-name/pa/paper-plane/tdlib.nix:1]: paper-tdlib and nixpkgs tdlib both use same versions of `gperf openssl readline zlib`
if say, we new openssl is incompatible, then both tdlib will need to be patched. Thats `an increase in maintenance costs.` 

as for license, if it changed, then in override license should also be overridden

I think only when there is sufficient differences between versions copy may be warranted. Currently there are no differences apart from src

Comment (Aleksanaa) [reply pkgs/by-name/pa/paper-plane/tdlib.nix]: I have changed to overrideAttrs, but if the build fails because the original package modifies the build script someday, I will not try to maintain this solution.

**Comment by kirillrdy** [comment]
@Aleksanaa thank you !

if it breaks, feel free to tag me, then we can have a look at possible ways of fixing it !

**Pull request merged**
